### PR TITLE
Add Cumsum double backwards support.

### PR DESCRIPTION
**Pull request by gchanan**
Also fix DivConstant for double backwards.
Remove unnecessary indexing for SelectionFunction.
Remove Repeat @once_differentiable decoration.

### CODE REVIEW

**Comment by gchanan** [comment]
As in PR https://github.com/pytorch/pytorch/pull/1753 I tested these using a modified gradgradcheck.

File: test/test_autograd.py [hunk 1]
```diff
@@ -1362,9 +1362,9 @@ class dont_convert(tuple):
     (Dot, (), ((L,), (L,)),),
     (Max, (), ((S, S, S),),),
     (Repeat, (), ((S, S, S, S), torch.Size([2, 3, 1, 2]))),
-    (Cumsum, (0,), ((S, S, S),)),
-    (Cumsum, (1,), ((S, S, S),), 'dim1'),
-    (Cumsum, (0,), ((S,),), '1d'),
+    (Cumsum, (), ((S, S, S), 0)),
+    (Cumsum, (), ((S, S, S), 1), 'dim1'),
+    (Cumsum, (), ((S,), 0), '1d'),
```
Comment (apaszke) [inline test/test_autograd.py]: can you add the negative dim checks here?

File: torch/autograd/_functions/tensor.py [hunk 2]
```diff
@@ -557,21 +556,19 @@ def backward(ctx, grad_output):
...
+        grad_input = torch.cumsum(-grad_output, dim=ctx.dim)
 
-        end_idx = grad_input.size(self.dim) - 1
-        grad_sum = grad_input.narrow(self.dim, end_idx, 1)
-        grad_input -= grad_sum.expand_as(grad_input)
-        grad_input += grad_output
-        return grad_input
+        end_idx = grad_input.size(ctx.dim) - 1
+        grad_sum = grad_input.narrow(ctx.dim, end_idx, 1)
+        grad_input = (grad_input - grad_sum.expand_as(grad_input)) + grad_output
```
Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: I think you should still be able to do the subtraction and addition in-place

Comment (gchanan) [inline torch/autograd/_functions/tensor.py]:
```
grad_input = (grad_input - grad_sum.expand_as(grad_input))
grad_input += grad_output
```
seems to work.  Make the first line inplace and running through gradgrad check gives me:
```
RuntimeError: in-place operations can be only used on variables that don't share storage with any other variables, but detected that there are 3 objects sharing it
```

Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: Right. I didn't notice that `grad_sum` shares storage with `grad_input`.

**Pull request merged**